server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java

```diff
@@ -53,10 +54,14 @@
 
     static final int TOTAL_PERMITS = Integer.MAX_VALUE;
     final Semaphore semaphore = new Semaphore(TOTAL_PERMITS, true); // fair to ensure a blocking thread is not starved
-    private final List<ActionListener<Releasable>> delayedOperations = new ArrayList<>(); // operations that are delayed
+    private final List<DelayedOperation> delayedOperations = new ArrayList<>(); // operations that are delayed
     private volatile boolean closed;
     private boolean delayed; // does not need to be volatile as all accesses are done under a lock on this
 
+    // only valid when assertions are enabled. Key is AtomicBoolean associated with each permit to ensure close once semantics. Value is an
+    // exception with some extra info in the message + a stack trace of the acquirer
+    private final Map<AtomicBoolean, RuntimeException> issuedPermits;
```

Comment:
any particular reason to use RuntimeException, and not just Throwable like in MockPageCacheRecycler

Reply:
Jason scared me into never using Throwable. will switch.